operator/crd.go

```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+)
+
+// CRDBackoff is the backoff for initial creation of CRDs by cilium-agent.
+var CRDBackoff = wait.Backoff{
+	Duration: 30 * time.Second,
+	Steps:    15,
+	Factor:   1.0,
+	Jitter:   0.1,
+}
+
+func getCRD(client clientset.Interface, CRDName string) error {
+	_, err := client.ApiextensionsV1beta1().CustomResourceDefinitions().Get(context.TODO(), CRDName, metav1.GetOptions{})
+	return err
+}
+
+func waitForCRDWithBackoff(client clientset.Interface, backoff wait.Backoff, CRDname string) error {
+	if err := retry.OnError(retry.DefaultBackoff, k8serrors.IsNotFound, func() error {
+		return getCRD(client, CRDname)
+	}); err != nil {
+		return errors.Wrapf(err, "timeout exceeded when waiting for CRD %s", CRDname)
+	}
+	return nil
```

Comment:
couldn't this be replaced with a cilium controller? It will also retry on error and we don't need to import more external packages.

Reply:
Cilium controller seems to have much more complex purpose than just simple "retry on error" action which I need here? I can try to think of the way of reusing parts of it, or rewriting it a bit in a way it will be easy to use for CRD watching.

Reply:
Or nevermind, sorry, I think I can use pkg/controller without any changes.